text/0085-api-keys.md

```diff
@@ -0,0 +1,463 @@
+- Title: API Keys
+- Start Date: 2021-10-15
+- Specification PR: [#85](https://github.com/meilisearch/specifications/pull/85)
+- Discovery Issue: [#51](https://github.com/meilisearch/product/issues/51)
+
+# API Keys
+
+## 1. Functional Specification
+
+### I. Summary
+
+Granular management of API keys is added to MeiliSearch. It is possible to restrict the access of an API key to certain actions on specific indexes.
+
+### II. Motivation
+
+To make MeiliSearch more reliable for teams, we extend the management and the possibilities of restrictions for the management of a MeiliSearch instance by introducing a concrete API resource (`API Key`). Security is a critical need, often tricky to negotiate as the stakes are high for a company.
+
+### III. Glossary
+
+| Term               | Definition |
+|--------------------|------------|
+| Main Key         | This is the main key that allows you to create other API keys. The main key is defined by the user when launching MeiliSearch, thus gives access to the `/keys` API endpoint. |
+| API Key            | API keys are stored and managed from the endpoint `/keys` by the main key holder. These are the keys used by the technical teams to interact with MeiliSearch at the level of the client code. |
+
+### IV. Personas
+
+| Persona | Role |
+|---------|------|
+| Anna    | Anna has an `ssh` access to the MeiliSearch instance and manages it on a daily basis. |
+| Mark    | Mark is a developer from the same company as Anna. He will implement the code to communicate with MeiliSearch to solve technical/product needs. |
+
+### V. `API Key` Explanations
+
+#### Summary Key Points
+
+- Occurrences of `master key` are changed by `main key`.
+- `MEILI_MASTER_KEY` environment variable is renamed `MEILI_MAIN_KEY`.
+- `--master-key` CLI option is renamed `--main-key`.
+- `/keys` management is restricted to the main key.
+- When a main key is set at MeiliSearch first-launch, we generate two pre-configured `API Key` resources. A `Default Search API Key` restricted to the search action and a `Default Admin API Key` to handle all operations (except managing API Keys) on MeiliSearch.
```

Comment:
What is the `expiresAt` value of the pre-configured keys? Are they any different from user-generated keys? Could a user change the `default search api key` to, say, include `document.add`?

Reply:
I will dedicate a part to its `Default API keys` generated by MeiliSearch.

These keys do not have an expiration date by default but can be modified at the convenience of the user. What is interesting is that Algolia for example in its system does not allow to modify this type of key.

They are mostly there to accelerate the implementation of the user's architecture like the public and private keys we had before.

In general, in a production environment, you must avoid having an API key with the search action that shares other actions because the search API key will most often be provided in the front-end code exposed to end-users and a malicious end-user could spoof it to perform unwanted operations on a MeiliSearch instance.